src/xcode/ENA/ENA/Source/Services/PPAnalyticsSubmitter/PPAnalyticsSubmitter.swift

```diff
@@ -0,0 +1,322 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+import OpenCombine
+
+protocol PPAnalyticsSubmitting {
+	/// Triggers the submission of all collected analytics data. Only if all checks success, the submission is done. Otherwise, the submission is aborted. In all cases, NO completion is called.
+	func triggerSubmitData()
+
+	#if !RELEASE
+	/// ONLY FOR TESTING. Triggers the submission of all collected analytics data. Only if all checks success, the submission is done. Otherwise, the submission is aborted. The completion calls are passed through to test the component.
+	func triggerSubmitData(ppacToken: PPACToken?, completion: ((Result<Void, PPASError>) -> Void)?)
+	#endif
+}
+
+final class PPAnalyticsSubmitter: PPAnalyticsSubmitting {
+
+	// MARK: - Init
+
+	init(
+		store: Store,
+		client: Client,
+		appConfig: AppConfigurationProviding
+	) {
+		self.store = store
+		self.client = client
+		self.configurationProvider = appConfig
+	}
+
+	// MARK: - Protocol PPAnalyticsSubmitting
+
+	func triggerSubmitData() {
+		// 0. Check if user has given his consent to collect data
+		if userDeclinedAnalyticsCollectionConsent {
+			Log.warning("Analytics submission abord due to missing users consent", log: .ppa)
+			return
+		}
+
+		configurationProvider.appConfiguration().sink { [weak self] configuration in
+
+			guard let self = self else {
+				Log.warning("Analytics submission abord due fail at creating strong self", log: .ppa)
+				return
+			}
+
+			// 1. Check configuration parameter
+			if Double.random(in: 0...1) > configuration.privacyPreservingAnalyticsParameters.common.probabilityToSubmit {
+				Log.warning("Analytics submission abord due to randomness", log: .ppa)
+				return
+			}
+
+			// 2. Last submission check
+			if self.submissionWithinLast23Hours {
+				Log.warning("Analytics submission abord due to submission last 23 hours", log: .ppa)
+				return
+			}
+
+			// 3a. Onboarding check
+			if self.onboardingCompletedWithinLast24Hours {
+				Log.warning("Analytics submission abord due to onboarding completed last 24 hours", log: .ppa)
+				return
+			}
+
+			// 3b. App Reset check
+			if self.appResetWithinLast24Hours {
+				Log.warning("Analytics submission abord due to app resetted last 24 hours", log: .ppa)
+				return
+			}
+
+			// 5. obtain usage data
+			let payload = self.obtainUsageData()
+
+			// 4. obtain authentication data
+			let deviceCheck = PPACDeviceCheck()
+			let ppacService = PPACService(store: self.store, deviceCheck: deviceCheck)
+
+			// 6. submit analytics data
+			ppacService.getPPACToken { [weak self] result in
+				switch result {
+				case let .success(token):
+					self?.submitData(with: token, for: payload, completion: nil)
+				case let .failure(error):
+					Log.error("Could not submit analytics data due to ppac authorization error", log: .ppa, error: error)
+					return
+				}
+			}
+
+		}.store(in: &subscriptions)
+	}
+
+	#if !RELEASE
+	func triggerSubmitData(
```

Comment:
If this is the same code as the one without the completion and the completion has the detault value nil. It can also be used in production code.

Reply:
aye, I will refactor this.